### ISSUE TITLE
qTox not using correct pixel format

### ISSUE THREAD

**Issue by seankhl** [qTox/qTox#2826]
Hello,

qTox: v1.2-238-g452c9e1, compiled from source
toxcore: newest as of yesterday
Qt: 5.5.1
OS version: Debian Stretch/Sid

One of my persistent problems with qTox has been that video4linux2 would change my framerate so something obnoxiously low whenever I upped the resolution (above 640x480, it would default to 10fps or below). I did some digging and found that it was defaulting to the yuyv format, which is uncompressed, so the fps at higher resolutions is hardware-limited! Instead, it should be using the mjpeg format, at _least_ for the higher resolutions, and perhaps in general (to lower the workload).

Hackishly, I achieved this behavior by adding this line:

```
av_dict_set(&options, "pixel_format", "mjpeg", 0);
```

at `cameradevice.cpp:169`.

This works, and now the setting options that I see in the A/V settings tab won't get overridden by v4l2. But now, swscale will complain about the `pix_fmt` that it ends up receiving (which for mjpeg is unfortunately the deprecated `yuvj422p`):

```
[swscaler @ 0x2d69c00] deprecated pixel format used, make sure you did set range correctly
```

Can we be sure that the range is set correctly? If not, any pointers for how to do so? If it's already set correctly, perhaps someone with more familiarity with the pull request norms here can add this option into the code (and perhaps in a less hackish way). Otherwise, if someone could help me get started on this range setting stuff, I can try to throw together a pull request myself.

This problem currently severely limits my video quality with using qTox and I'm afraid it probably affects most linux users too. It would be a low hanging fruit for a big improvement.

**Comment by tux3** [comment]
Thanks, I took your solution and silenced the useless libswscale warning spam, this should have a positive effect for the vast majority of webcams.

The fix for the warning was just looking at the [swscale source](https://git.videolan.org/?p=ffmpeg.git;a=blob;f=libswscale/utils.c;h=74f0bd321eba33ba8f3a669d506c569c781d2bb8;hb=HEAD) and satisfying the check line 1152.
It doesn't actually care that the range is set correctly (and it will fix the range itself anyways), but it yells if it sees deprecated pixel formats.

**Comment by seankhl** [comment]
> It doesn't actually care that the range is set correctly (and it will fix the range itself anyways), but it yells if it sees deprecated pixel formats.

I sure hope you're right about this! :^p

Thanks. Another thing to note is that some webcams might have other format options, in particular, h264 might be natively supported (I've seen some webcams that can do this). I'm struggling to find a list of possible formats, but it seems best to default to h264 before mjpeg if possible, too...

**Comment by tux3** [comment]
A cleaner way to do this would be to query the list of available pixel formats and pick the better one, you could try to hack on this if you want, or open another wishlist issue.

**Comment by seankhl** [comment]
Yes yes, querying the camera is definitely best. I just find the ffmpeg documentation so painful to sift though...

I'll see if I can't throw something together.

**Comment by seankhl** [comment]
Hey I saw the problem and I've been working on the code. I am going to finish later tonight probably, but I have a way better method for doing this than the above that should work in the general case.

The one problem is that the name of these pixel formats in string form (`char *`) is not documented anywhere and there seem to be no conversion function. All I've found is this:

https://www.ffmpeg.org/doxygen/0.6/pixfmt_8h-source.html

Of course, the AVDictionary only sets the pixel format based on a string -_-. So any info about how to do this in the general case would be great.

_edit_ So it seems like these pixel formats don't correspond to the set of possibilities for the AVDictionary. Since, well, there's no mjpeg format listed there...

_edit 2_ Looks like we might want to use coded IDs. See https://www.ffmpeg.org/doxygen/2.7/v4l2-common_8c_source.html
